infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java

```diff
@@ -324,46 +360,76 @@ public static ReplicationSubscriber getDefaultSubscriber() {
                 .build();
     }
 
-    public void stopReplication() {
+    /**
+     * Stop all sessions
+     */
+    public void stopSessions() {
         log.info("Stopping log replication.");
-        replicationManager.stop();
+        stopSessions(sessions);
     }
 
-    private void stopReplication(Set<LogReplicationSession> sessions) {
+    /**
+     * Stop replication for a set of sessions and clear in-memory info
+     * @param sessions
+     */
+    private void stopSessions(Set<LogReplicationSession> sessions) {
+        if (sessions.isEmpty()) {
+            return;
+        }
         if (replicationManager != null) {
-            replicationManager.stop(sessions);
+            // stop replication fsm and source router for outgoing sessions
+            replicationManager.stop(sessions.stream().filter(outgoingSessions::contains).collect(Collectors.toSet()));
+            router.stop(sessions);
         }
 
         sessions.forEach(session -> {
             this.sessions.remove(session);
             this.outgoingSessions.remove(session);
+            this.incomingSessions.remove(session);
         });
 
     }
 
-    public synchronized void startReplication() {
-
-        log.info("Start replication for all sessions count={}", outgoingSessions.size());
+    /**
+     * Connect to remote cluster for sessions that are connection initiator.
+     * If local cluster is SOURCE for any session: create FSMs and initiate connection to remote sinks
+     * If local cluster is SINK for any session: initiate connection to remote sources.
+     */
+    public synchronized void connectToRemoteClusters() {
+        router.createTransportClientAdapter(configManager.getServerContext().getPluginConfigFilePath());
+
+        newSessionsDiscovered.stream()
+                .filter(session -> topology.getRemoteClusterEndpoints().containsKey(session.getSourceClusterId()) ||
+                        topology.getRemoteClusterEndpoints().containsKey(session.getSinkClusterId()))
+                .forEach(session -> {
+                    if (outgoingSessions.contains(session)) {
+                        // initiate connection to remote Sink cluster
+                        router.connect(topology.getAllClustersInTopology().get(session.getSinkClusterId()), session);
+                    } else {
+                        // initiate connection to remote Source cluster
+                        router.connect(topology.getAllClustersInTopology().get(session.getSourceClusterId()), session);
```

Comment:
I was thinking it might become too verbose. 
Now that we don't have to distinguish different kinds of router, the object name can be just "router"...?